docs/spec/auth/subkeys_proposal.md

```diff
@@ -0,0 +1,156 @@
+
+```golang
+type SubKeyMetadata struct {
+  PubKey               sdk.AccPubKey
+  PermissionedRoutes   []string
+  DailyFeeAllowance    sdk.Coins
+  DailyFeeUsed         sdk.Coins
+}
+```
+
+We create a new `Account` type called `SubKeyAccount` that stores a list of `SubKeyMetadata`s.
+
+```golang
+type SubKeyAccount struct {
+  Address       AccAddress
+  Coins         Coins
+  PubKey        sdk.AccPubKey
+  AccountNumber uint64
+  Sequence      uint64
+  SubKeys       []SubKeyMetadata
+}
+```
+
+In the `StdSignature` type we need to add a new field called `PubKeyIndex` which tells the anteHandler which SubKey to use to verify this signature.
+
+```golang
+// StdSignature represents a sig
+type StdSignature struct {
+	crypto.PubKey                  `json:"pub_key"` // optional
+  Signature     []byte           `json:"signature"`
+  PubKeyIndex   uint
+}
+```
+
+If a `StdSignature.PubKeyIndex` is 0, this means that the account's "master" PubKey should be used to verify the signature.  If the `StdSignature.PubKeyIndex` is `> 0`, it verifies the signature using the corresponding **1-indexed** SubKey in the SubKeyAccount.SubKeys slice.  Note, the SubKeys slice is 1-indexed rather than 0-indexed because the 0 index is used to refer to the master PubKey.  See pseudocode below:
+
+```
+if stdSig.PubKeyIndex == 0 {
+  verify stdSig using acc.PubKey
+} else {
+  verify stdSig using acc.SubKeys[stdSig.PubKeyIndex - 1]
+}
+```
+
+In the AnteHandler, if a transaction comes from a SubKey, it verifies the msg route is permitted.
+
+```
+if msg.Route not in subKeyMetadata.PermissionedRoutes {
+  return ErrNotPermitted
+}
+```
+
+When a transaction signed by a SubKey is accepted and has paid fees, we need to log the fees it has paid.  We increase the `SubKeyMetadata.DailyFeeUsed` field.
+
+If a new transaction from a subkey comes in and the `tx.Fee + subKeyMetadata.DailyFeeUsed > subKeyMetadata.DailyFeeAllowance`, then the transaction is rejected as the subkey is trying to exceed its daily fee allowance.
+
+But now we need a way to decrease the `DailyFeeUsed` field once transactions are past the 24 hour window.  To do this we us a time based iterator similar to the ones used in the governance and staking queues.
+
+We start by defining a new struct called a `DailyFeeSpend`, which is what will be inserted into the Queue.
+
+```golang
+type DailyFeeSpend struct {
+  Address              sdk.AccAddress
+  SubKeyIndex          uint
+  FeeSpent             sdk.Coins
+}
+```
+
+When a transaction using a SubKey has paid its fee, we insert a new entry into this queue's store.
+- The key is `sdk.FormatTimeBytes(ctx.BlockTime.Add(time.Day * 1)`
+- The value is the Amino marshalled `DailyFeeSpend`
+
+Finally in the EndBlocker we iterate over the all the DailyFeeSpends in the queue which are expired (more than a day old).  We prune them from state, and deduct their FeeSpent amount from the corresponding SubKeyMetadata.
```

Comment:
Let me just confirm this idea - the queue is organized such that if nothing is expired, the queue will only need to check the top record of the queue each endblock to confirm this correct?

Reply:
Yep.  The keys are the `sdk.SortableTimeFormat` bytes, and so the EndBlocker just asks for an iterator from `nil` to `sdk.FormatTimeBytes(ctx.BlockHeader().Time)`.  If there's nothing that is expired, the iterator will just come back empty, and it will go on.  Won't have to even do any Amino unmarshals.

Reply:
nice, thanks